Factored out duration parsing from car_simulation.h's CreateRigidBody…

liangfok opened the pull request:
It doesn't really make sense to parse the duration in method `CreateRigidBodySystem()`. The increased efficiency of not having to iterate through the command line arguments twice is negligible. Thus, I believe this change increases code clarity.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/3015)
<!-- Reviewable:end -->

liangfok commented:
+@jwnimmer-tri for feature review.
+@sherm1 for platform review.

Thanks!

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-:-KO5tpXE3I9ZvTumuGSv:-1069660197)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@drake-jenkins-bot linux-gcc-experimental-ros please

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-:-KO64SL5MMyItiJQS_qS:-206704113)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 5 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[CHANGELOG.md, line 26 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO67Xza8BGPlrx_lyqq:-KO67Xzbyx9wIqgEeFnc:659410940) ([raw file](https://github.com/robotlocomotion/drake/blob/fb481e8b72458072187c983837ad8d75a3a9b5b5/CHANGELOG.md#L26)):*
> ```Markdown
> ### Changed
> 
>  - [#3015][] Removed `duration` parameter from CreateRigidBodySystem() in car_simulation.h
> ```

Line too long (80).

BTW I am not sure that this entry is really needed, since part of examples, not Drake.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-:-KO69Ii8TEHY2rN4w3pF:1417088721)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@jwnimmer-tri I addressed your comments. 

Now waiting on CI and platform review from @sherm1.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[CHANGELOG.md, line 26 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO67Xza8BGPlrx_lyqq:-KO69k4RvGfp6IiClEf6:216517915) ([raw file](https://github.com/robotlocomotion/drake/blob/fb481e8b72458072187c983837ad8d75a3a9b5b5/CHANGELOG.md#L26)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Line too long (80).
> 
> BTW I am not sure that this entry is really needed, since part of examples, not Drake.
</details>
Done. I removed the entry.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-:-KO69mBYyjNsAcc5BXRX:1022554567)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@drake-jenkins-bot linux-gcc-experimental-ros please

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-:-KO69slSQFZpj8VLco1R:-206704113)*
<!-- Sent from Reviewable.io -->

liangfok commented:
CI failure most likely due to a Github.com hiccup. 

@drake-jenkins-bot retest this please.

liangfok commented:
CI passed on second attempt. Thus, now waiting on @sherm1 for platform review.

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-:-KO6pJTQ4Tnm041x1elA:-958235585)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Platform review: <img class="emoji" title=":lgtm_cancel:" alt=":lgtm_cancel:" align="absmiddle" src="https://reviewable.io/lgtm_cancel.png" height="20" width="61"/>. See my comments below. This PR doesn't achieve its stated goal of improving clarity. If you want to do that you should separate command line parsing from the consumers of the command line arguments. It doesn't help to parse in separate places if each place has to know every other place's business in order to find their own arguments.

---

Reviewed 3 of 5 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[drake/examples/Cars/car_simulation.cc, line 23 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7FGMwJE1IJx0uBbpw:-KO7FGMwJE1IJx0uBbpx:-540121422) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L23)):*
> ```cpp
>                                                        const char* argv[]) {
>   if (argc < 2) {
>     std::cerr << "Usage: " << argv[0] << " vehicle_model [world sdf files ...]"
> ```

I'm not sure what this usage message is trying to say in the `[ ]`. What is a "world sdf file"? Or does that mean 1. world, then 2. an sdf, then one or more files? Does this mean I can have more than one world? Is this actually restricted to only sdf files (the code below looks like it would take other formats too). Can you clarify the message?

---

*[drake/examples/Cars/car_simulation.cc, line 24 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7EygMzig4Xls6Qky4:-KO7EygMzig4Xls6Qky5:423364308) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L24)):*
> ```cpp
>   if (argc < 2) {
>     std::cerr << "Usage: " << argv[0] << " vehicle_model [world sdf files ...]"
>               << " --duration [duration in seconds]" << std::endl;
> ```

Since `CreateRigidBodySystem()` is no longer responsible for all the parsing, it seems wrong to have it issue the usage message, which includes the `--duration` flag which is not parsed here.  If the duration were malformed, you would have to issue the same message again from whatever code discerned that problem.

I suggest that you factor out a Usage() message function that issues the appropriate message and then exits. That could be called from any of the methods that run through (argc,argv) and might find something amiss.

---

*[drake/examples/Cars/car_simulation.cc, line 35 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7DoP43wNVaojPQZBA:-KO7DoP43wNVaojPQZBB:-1686605978) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L35)):*
> ```cpp
>   // Adds the environment to the rigid body tree.
>   for (int ii = 2; ii < argc; ++ii) {
>     if (std::string(argv[ii]) != "--duration") {
> ```

Wait, what?? Here `CreateRigidBodySystem()` still knows about all possible flags? That doesn't make sense. Either this method should do all the parsing or it should parse things of local relevance only. I think what you want here is to find all the sdf files, which I believe would be any remaining argument not beginning with `--`.

This needs improved comments also.

---

*[drake/examples/Cars/car_simulation.cc, line 36 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7FvU5rY-cioq9ekHE:-KO7FvU5rY-cioq9ekHF:672252671) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L36)):*
> ```cpp
>   for (int ii = 2; ii < argc; ++ii) {
>     if (std::string(argv[ii]) != "--duration") {
>       rigid_body_sys->addRobotFromFile(argv[ii], DrakeJoint::FIXED);
> ```

BTW, seems to allow more than sdf?

---

*[drake/examples/Cars/car_simulation.cc, line 38 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7G4gyNAND2cOAtLFe:-KO7G4gyNAND2cOAtLFf:-795629136) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L38)):*
> ```cpp
>       rigid_body_sys->addRobotFromFile(argv[ii], DrakeJoint::FIXED);
>     } else {
>       ++ii;  // Skips the value immediately after "--duration".
> ```

Oops, now this code has to know whether the flag takes zero or one argument. This seems like information leakage that defeats the purpose of this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-:-KO7Gheo0iaYDAqVUv5G:1990028304)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/Cars/car_simulation.cc, line 35 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7DoP43wNVaojPQZBA:-KO7gOQkLvyG87-RWorX:1940450428) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L35)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Wait, what?? Here `CreateRigidBodySystem()` still knows about all possible flags? That doesn't make sense. Either this method should do all the parsing or it should parse things of local relevance only. I think what you want here is to find all the sdf files, which I believe would be any remaining argument not beginning with `--`.
> 
> This needs improved comments also.
</details>
FWIW -- I was comfortable with the coupling / brittleness between these different methods (in terms of what arguments are supported, and how many arguments the flags take), because these methods are all defined in the same `.cc` file, and limited in scope to just the Cars example(s).

Until we add `Gflags` to the build and do proper command-line parsing, it is going to suck, and this PR seemed like a strict improvement even if the status quo is fairly awful.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/Cars/car_simulation.cc, line 35 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7DoP43wNVaojPQZBA:-KO8C6fJDRIIGYuahLRE:1730302763) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L35)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> FWIW -- I was comfortable with the coupling / brittleness between these different methods (in terms of what arguments are supported, and how many arguments the flags take), because these methods are all defined in the same `.cc` file, and limited in scope to just the Cars example(s).
> 
> Until we add `Gflags` to the build and do proper command-line parsing, it is going to suck, and this PR seemed like a strict improvement even if the status quo is fairly awful.
> 
</details>
Ah, now you're talking! Adding [gflags](https://github.com/gflags/gflags) to Drake's externals seems like a very worthwhile project. What about doing that instead, @liangfok? After that you could revisit this PR and make the code significantly better, as well as set a good example for how other people should be handling command lines.

I still don't see the present PR as an improvement.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@sherm1 I responded to your comments. Here's what the new usage instructions look like on my OS X machine:

![Screen Shot 2016-08-01 at 11.23.24 PM.png](https://files.reviewableusercontent.io/images/reviewable/1388098/WDK0JRJ_aONhAjgcm5m_rDjv/Screen%20Shot%202016-08-01%20at%2011.23.24%20PM.png)

PTAL. Thanks!

---

Review status: 2 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/Cars/car_simulation.cc, line 23 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7FGMwJE1IJx0uBbpw:-KO8AWtcsXvUx1EQqheE:-1268516567) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L23)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> I'm not sure what this usage message is trying to say in the `[ ]`. What is a "world sdf file"? Or does that mean 1. world, then 2. an sdf, then one or more files? Does this mean I can have more than one world? Is this actually restricted to only sdf files (the code below looks like it would take other formats too). Can you clarify the message?
</details>
I moved the message print statement into its own method and clarified it.

---

*[drake/examples/Cars/car_simulation.cc, line 24 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7EygMzig4Xls6Qky4:-KO8B1T7TxR7lhhKXQ8X:-1158607847) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L24)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Since `CreateRigidBodySystem()` is no longer responsible for all the parsing, it seems wrong to have it issue the usage message, which includes the `--duration` flag which is not parsed here.  If the duration were malformed, you would have to issue the same message again from whatever code discerned that problem.
> 
> I suggest that you factor out a Usage() message function that issues the appropriate message and then exits. That could be called from any of the methods that run through (argc,argv) and might find something amiss.
</details>
I factored out the usage print statement into its own method but decided to let the calling context decide whether to exit with an error code or throw an exception.

---

*[drake/examples/Cars/car_simulation.cc, line 35 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7DoP43wNVaojPQZBA:-KO8BD_mCSGt5bcOYNgt:-1051166565) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L35)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> FWIW -- I was comfortable with the coupling / brittleness between these different methods (in terms of what arguments are supported, and how many arguments the flags take), because these methods are all defined in the same `.cc` file, and limited in scope to just the Cars example(s).
> 
> Until we add `Gflags` to the build and do proper command-line parsing, it is going to suck, and this PR seemed like a strict improvement even if the status quo is fairly awful.
> 
</details>
The logic is a slightly more complex than simply ignoring arguments that begin with `--` because once `--duration` is found, the subsequent value should also be ignored. Since this logic existed before this PR, I'd like to keep it here. I agree that factoring out all parsing into a separate chunk of code would be ideal, but that's inherited technical debt that I'd prefer to pass on to a future PR.

---

*[drake/examples/Cars/car_simulation.cc, line 36 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7FvU5rY-cioq9ekHE:-KO7ucmAh-HSdd0Ld91h:55978546) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L36)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, seems to allow more than sdf?
</details>
Yes, it supports both SDF and URDF. I've updated the description.

---

*[drake/examples/Cars/car_simulation.cc, line 38 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7G4gyNAND2cOAtLFe:-KO8CFy4-sVYwyTfG3Y7:1594228654) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L38)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Oops, now this code has to know whether the flag takes zero or one argument. This seems like information leakage that defeats the purpose of this PR.
</details>
With the newly documented usage instructions (see: `PrintUsageInstructions()` above), it should now be clear that the a floating point value should reside after the `--duration` flag.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-:-KO8CoFGLI_zxybabP9o:-993196939)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@drake-jenkins-bot linux-gcc-experimental-ros please

---

Review status: 2 of 4 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-:-KO8H73s-Yp9-P8VlfZB:-206704113)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 2 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/examples/Cars/car_simulation.cc, line 35 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7DoP43wNVaojPQZBA:-KO8I5IvNT9Eu4GDpA3I:356920408) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L35)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> The logic is a slightly more complex than simply ignoring arguments that begin with `--` because once `--duration` is found, the subsequent value should also be ignored. Since this logic existed before this PR, I'd like to keep it here. I agree that factoring out all parsing into a separate chunk of code would be ideal, but that's inherited technical debt that I'd prefer to pass on to a future PR.
</details>
I'm afraid of making this PR too big if I include the addition of `gflags` and code modifications that make use of it in this PR. How about let's push it to a future PR? See: #3027.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 2 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/examples/Cars/car_simulation.cc, line 38 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7G4gyNAND2cOAtLFe:-KO8McNXxjKxJTNQszr6:1165085656) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L38)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> With the newly documented usage instructions (see: `PrintUsageInstructions()` above), it should now be clear that the a floating point value should reside after the `--duration` flag.
</details>
I replaced the `++ii` with a `break`. Is this any better?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@drake-jenkins-bot linux-gcc-experimental-ros please

---

Review status: 2 of 4 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-:-KO8MtkvYizdgZvFm6WM:-206704113)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Usage looks great (one question). See my long comment below about how I think you should proceed with this PR.

---

Reviewed 1 of 3 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/examples/Cars/car_simulation.cc, line 23 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7FGMwJE1IJx0uBbpw:-KO8QM3LjuJa8R_PNyjq:1506377092) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L23)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I moved the message print statement into its own method and clarified it.
</details>
That's a very nice message! I noticed that you changed `vehicle_model` to `[vehicle model file]`. In this context, brackets mean "optional" where before it was required. Which is correct?

---

*[drake/examples/Cars/car_simulation.cc, line 35 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7DoP43wNVaojPQZBA:-KO8Tqunt6-O1RazZATF:-922247271) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L35)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I'm afraid of making this PR too big if I include the addition of `gflags` and code modifications that make use of it in this PR. How about let's push it to a future PR? See: #3027.
</details>
Agreed. See my comment below for what I think you should do.

---

*[drake/examples/Cars/car_simulation.cc, line 38 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7G4gyNAND2cOAtLFe:-KO8RMqFzR8dpaOl7mLa:1093099941) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L38)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I replaced the `++ii` with a `break`. Is this any better?
</details>
No, sorry that doesn't help and neither does replacing "--duration" with a defined constant. The problem is that you wanted to separate out the flag parsing from the argument parsing (which would be good), but you still had to leave CreateRigidBodySystem() (CRBS) with a full understanding of the flag, even though it has no use for that flag. So you didn't actually separate the two parsing tasks after all. Since that was the purpose of the PR, it doesn't make sense to merge it like this. It 
makes the whole thing *more* complicated and entangled rather than less. Using `gflags` or `getopt` you would separately parse the command line to extract flags, then CRBS would deal only with the arguments, and some other piece of code would deal only with the `duration` flag which would achieve the original goal of this PR very nicely.

I do agree you should not mix in the `gflags` project with this PR. Here is what I think you should do:
- Keep the very nice "usage" message you added, as well as other documentation improvements you made.
- Revert to the old code where CRBS parses the complete command line and returns the duration value.
- Merge this PR like that.
- Then, get gflags into Drake.
- File a new PR that uses gflags in car_simulation.
- Watch for an enthusiastic "L G T M" from me!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-:-KO8U-nNuorSRgnO2IH2:1699773585)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 3 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[drake/examples/Cars/car_simulation.h, line 33 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KOA-BMRb1F_alXIn9Op:-KOA-BMScUXdC4ttDAwY:-84186818) ([raw file](https://github.com/robotlocomotion/drake/blob/b7cf676a0ea6d455e1c128d04ffea5db3da722e4/drake/examples/Cars/car_simulation.h#L33)):*
> ```objc
>  */
> DRAKECARS_EXPORT
> void PrintUsageInstructions(const std::string executable_name);
> ```

BTW Prefer pass-by-reference `const std::string& executable_name`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks.

@drake-jenkins-bot linux-gcc-experimental-ros please

---

Review status: 2 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/examples/Cars/car_simulation.cc, line 23 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7FGMwJE1IJx0uBbpw:-KOA2HSuZ_AZ6ydy6GNT:-446030081) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L23)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> That's a very nice message! I noticed that you changed `vehicle_model` to `[vehicle model file]`. In this context, brackets mean "optional" where before it was required. Which is correct?
</details>
I modified the usage statement to clarify which parts are optional and which are required. The arguments in square brackets are effectively macro replacements. Here's what the updated text looks like:

![Screen Shot 2016-08-02 at 7.47.23 AM.png](https://files.reviewableusercontent.io/images/reviewable/1388098/FxQHHwoFzxXgBAxveTQQwBh7/Screen%20Shot%202016-08-02%20at%207.47.23%20AM.png)

---

*[drake/examples/Cars/car_simulation.cc, line 35 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7DoP43wNVaojPQZBA:-KOA4MozsUytXyVscp9m:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L35)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Agreed. See my comment below for what I think you should do.
</details>
Done.

---

*[drake/examples/Cars/car_simulation.cc, line 38 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7G4gyNAND2cOAtLFe:-KOA5zh2lajnZ-RRxcDW:-1839196547) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L38)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> No, sorry that doesn't help and neither does replacing "--duration" with a defined constant. The problem is that you wanted to separate out the flag parsing from the argument parsing (which would be good), but you still had to leave CreateRigidBodySystem() (CRBS) with a full understanding of the flag, even though it has no use for that flag. So you didn't actually separate the two parsing tasks after all. Since that was the purpose of the PR, it doesn't make sense to merge it like this. It 
> makes the whole thing *more* complicated and entangled rather than less. Using `gflags` or `getopt` you would separately parse the command line to extract flags, then CRBS would deal only with the arguments, and some other piece of code would deal only with the `duration` flag which would achieve the original goal of this PR very nicely.
> 
> I do agree you should not mix in the `gflags` project with this PR. Here is what I think you should do:
> - Keep the very nice "usage" message you added, as well as other documentation improvements you made.
> - Revert to the old code where CRBS parses the complete command line and returns the duration value.
> - Merge this PR like that.
> - Then, get gflags into Drake.
> - File a new PR that uses gflags in car_simulation.
> - Watch for an enthusiastic "L G T M" from me!
</details>
Done. I reverted to the previous API where there was a `duration` parameter. However, I also made this parameter optional by giving it a default value of `nullptr`.

---

*[drake/examples/Cars/car_simulation.h, line 33 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KOA-BMRb1F_alXIn9Op:-KOA4FFcv4i0d8Lf-Spa:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/b7cf676a0ea6d455e1c128d04ffea5db3da722e4/drake/examples/Cars/car_simulation.h#L33)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Prefer pass-by-reference `const std::string& executable_name`.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-:-KOA6Yo1ru3myKsNA77A:1082872181)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@drake-jenkins-bot linux-gcc-experimental-ors please

---

Review status: 1 of 4 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-:-KOA6sZNoSV914NOJOCO:-1606513931)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@drake-jenkins-bot linux-gcc-experimental-ros please

---

Review status: 1 of 4 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-:-KOA6vo5mGgXHSYo-rNj:-206704113)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 2 files at r6, 1 of 2 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015)*
<!-- Sent from Reviewable.io -->

liangfok commented:
This PR passed all CI tests and is thus ready for a platform review from @sherm1.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-:-KOAe6ND3Wd2s4qFS5lp:-1927223164)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review pass done. Liang, Sorry, I'm still not happy with this. The net effect of these changes is that the code (a) is more complicated than it was before (for example, now treats the duration argument as optional adding more code paths, then has to parse anyway to dodge the argument, then doesn't return the parsed value since it only uses the duration==nullptr case), (b) now parses the argument list twice in different places, and (c) now contains a new parsing method (ParseDuration) that will be obsoleted by issue #3027 (gflags). I don't see that you have gained anything with these changes. 

---

Reviewed 2 of 2 files at r6, 1 of 2 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Cars/car_simulation.cc, line 23 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7FGMwJE1IJx0uBbpw:-KOBuWqut_ObEkXYnAg5:1412946909) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L23)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I modified the usage statement to clarify which parts are optional and which are required. The arguments in square brackets are effectively macro replacements. Here's what the updated text looks like:
> 
> ![Screen Shot 2016-08-02 at 7.47.23 AM.png](https://files.reviewableusercontent.io/images/reviewable/1388098/FxQHHwoFzxXgBAxveTQQwBh7/Screen%20Shot%202016-08-02%20at%207.47.23%20AM.png)
> 
> 
</details>
Square brackets in a command line description always means "optional" (by long-standing convention). Most readers will look at the first line and not read all the text, so will (reasonably) assume that the "vehicle model file" argument is optional. You shouldn't put it in square brackets unless it is actually optional. I would suggest just writing it as `vehicle_model_file` with no brackets.

---

*[drake/examples/Cars/car_simulation.h, line 52 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KOBvSxV5Vp2nlqIjsnr:-KOBvSxV5Vp2nlqIjsns:1404211738) ([raw file](https://github.com/robotlocomotion/drake/blob/cd30edebbb7acea9bedb62b16f4205b8b9db28b5/drake/examples/Cars/car_simulation.h#L52)):*
> ```objc
>  *
>  * @param[out] duration The duration over which the simulation should run. The
>  * simulation runs from time zero seconds to time \p duration seconds. If no
> ```

BTW, I think we agreed on using `@` rather than `\` for doxygen commands.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-:-KOBwQsXAuszvUMzamzL:-1347556362)*
<!-- Sent from Reviewable.io -->

liangfok commented:
FTR, the `ParseDuration()` method existed prior to this PR:

https://github.com/RobotLocomotion/drake/blob/master/drake/examples/Cars/car_simulation.h#L65

I'll proceed to make the duration parameter required.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-:-KOC-6yiFZZV5wi3ADIa:-144526437)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
> FTR, the ParseDuration() method existed prior to this PR

Oh, I'm sorry -- I didn't realize that. What was it doing if CreateRigidBodySystem() already parsed duration?

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-:-KOC2EIG1HD-GDALSe_l:1919216164)*
<!-- Sent from Reviewable.io -->

liangfok commented:
No worries. It was being used by [drake-distro/drake/examples/Cars/multi_car_sim_lcm.cc](https://github.com/RobotLocomotion/drake/blob/master/drake/examples/Cars/multi_car_sim_lcm.cc#L64) which hard-codes the SDF file but still needs to parse the `--duration`` flag.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/Cars/car_simulation.cc, line 23 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7FGMwJE1IJx0uBbpw:-KOC1306LTWnrrOfxsjs:-1246255366) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L23)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Square brackets in a command line description always means "optional" (by long-standing convention). Most readers will look at the first line and not read all the text, so will (reasonably) assume that the "vehicle model file" argument is optional. You shouldn't put it in square brackets unless it is actually optional. I would suggest just writing it as `vehicle_model_file` with no brackets.
</details>
Done. Here is the updated text:

---

*[drake/examples/Cars/car_simulation.h, line 52 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KOBvSxV5Vp2nlqIjsnr:-KOC1UA01fH7Lr2J_bSV:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/cd30edebbb7acea9bedb62b16f4205b8b9db28b5/drake/examples/Cars/car_simulation.h#L52)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, I think we agreed on using `@` rather than `\` for doxygen commands.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-:-KOC4TiXyud4cOOL1env:1846920077)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I've addressed all reviewer comments. PTAL.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Cars/car_simulation.cc, line 23 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7FGMwJE1IJx0uBbpw:-KOC5JFPNH6lWXT_JgmV:1431686587) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L23)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Done. Here is the updated text:
</details>
Oops, here's the screenshot:


![Screenshot from 2016-08-02 17:11:32.png](https://files.reviewableusercontent.io/images/reviewable/1388098/zQeB49uFn6aKgM7lmugMDxA9/Screenshot%20from%202016-08-02%2017%3A11%3A32.png)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-:-KOC5MM0UcDioRoUecjr:616398360)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 4 of 4 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>. Thanks for your patience, @liangfok. I spotted one typo; comment is marked "BTW" to be non blocking but should get repaired before merging.

---

Reviewed 4 of 4 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Cars/car_simulation.cc, line 23 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KO7FGMwJE1IJx0uBbpw:-KOC8kyzSUcjYyO8Rdqp:1566170132) ([raw file](https://github.com/robotlocomotion/drake/blob/ef62b1b1e67ec06afaf8ee59d2ecf1bedd6ff519/drake/examples/Cars/car_simulation.cc#L23)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Oops, here's the screenshot:
> 
> 
> ![Screenshot from 2016-08-02 17:11:32.png](https://files.reviewableusercontent.io/images/reviewable/1388098/zQeB49uFn6aKgM7lmugMDxA9/Screenshot%20from%202016-08-02%2017%3A11%3A32.png)
> 
</details>
OK, looks great -- I like those changes. No spaces -> no ambiguity!

---

*[drake/examples/Cars/car_simulation.cc, line 48 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-KOC8cZr-WmBxdRYYf5f:-KOC8cZr-WmBxdRYYf5g:1507327469) ([raw file](https://github.com/robotlocomotion/drake/blob/e46b03c19fa4788c74b5599d2498e612efebc7c0/drake/examples/Cars/car_simulation.cc#L48)):*
> ```cpp
>     << "  - duration_in_seconds is the number of seconds (floating point) to"
>     << std::endl
>     << "    to run the simulation. This value is in simulation time."
> ```

BTW, this says "to to run ...".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015#-:-KOC9I3hbLo3BM8rzHvq:659993399)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3015)*
<!-- Sent from Reviewable.io -->
